### PR TITLE
feat: react-combobox space conditionally inserts character when freeform is true

### LINKED ISSUE #26295: [Bug]: Cannot type space in a Combobox when freefrom is enabled
### Library

React Components / v9 (@fluentui/react-components)

### System Info

```shell
Tested on `@fluent/react-components==9.9.1`, Microsoft Edge 107.0.1418.26 on Windows 11
```

### Are you reporting Accessibility issue?

no

### Reproduction

[https://codesandbox.io/api/v1/sandboxes/define?parameters=N4IgZg[...]sCc6DQA&query=file%3D%2Fexample.tsx](https://codesandbox.io/api/v1/sandboxes/define?parameters=N4IgZglgNgpgziAXKGAPAhgWwA6wHQAucqSoEcAQhAHboBOAnkmOlHDADQgDGA9tQRgCkICDl50CAAmBTM6ANYwAygQaw4HKXAAWEgjvTUAJpqkBXdgEljUgL5SwdXpikAdEAAEwUc0ILmEAD0dDDo3AQAtHzi1P5wHgDcbtRi2PoyUgDCLgBGvPmoWjmY-YUACs7YZgDy2AQQ_PaOzq4e3r7-gSFhEdEu6XECcEHm1HAE6LmwSSlpGQBUUuhwUgBKvdJOLlIA5KHhBLvJ1Hzj0pYqahpSALxyilfq8AAUwClSUs68BIgyH58pEEglJVOEFFIDDApFApjAoMt8gA3aFQxwQeG2ADuEAMyykAHN0NgAZ9jORcOgGH9dgS6BBjLsOKTCfTjAAVGA4WGCNa8LFwGmhbBhAgvACMTgAlEyWQArSwNMAMKyCTCCvYTehHZnUQFSPCG3T6QwmOB4InYF67ABM2FQuyluv18lQAHUGQYaQAWAAMvvtuwBdhSdilJzQ6UkUjOEykADFQjAwBJXPcXtgqhrytqIKwADwlMq8VCVXjVAB8UruFf-epj_DjMXyNjuFmsxmtzd4kWMyfQ5igR3DANj0nLDUbbYA2rssugdXt54I6NhoLC6EylwvILot3OdDAYJMdPusjoINwlNQz7wrwuMVBb1j9wARXgEt-8KDYC83rS7PGMB0KEi6ARAoQ-Aw-7xuQp4AfGJYwSWUiciBGKbghzifgBAASWATMB-7KLQSi7AAuicnxjlI07yAQ3B_gSdSTuMWjsAQACyC6MTQzH1I04zkW2GyHHglxgoILzToaeAToJcDkSO9Y0dO3CKi4yhhHQjHsceWQaZgWn0Ixwn3KJETiewkkwPmEz0tQBJSAAPhYJjJjQMDGBWLzKdRjbSBMzyrPcEnXK8fkNucUj8OeRgEjAfxFgUJZltUs6xaaCUUW2MAogINZ1vqNFIqwfi5flBCEPQCVVaVnSEPSmC-VRgI0fRjHwG28mNngkBDsBLw9Xqty1sNhC8AAMvywHzuwvl4DQfaoDUYAvPVfgTdNWKzSsMC-dWtxHVIvqRZ8HHcQxTEsQpLwdYecBnVIEBgFI61lTAeCwI5eIAGS_Q8V3wF9QgEni-ZSOK1bvPW-ocQZEyadpjHvZ0T0OPC7BFfq2j6YZxk6ToLxjH2kBxMY6PBq1NH8Dd_BabAERJXkKWllmGXUHT1AMzAEQ5emeX-FoxgLugh21jDxUBYDvGOVzbYi5MckCfwnKoNI_0xSr4yLacvh9nALyK-gyusWrBBPS9b33XxXPQyy5144jRnI0TJMeeT6NSJj0KSzj8P467Rui6bgnm5T9Z2K1oHmHQer5uSSIxrCcBwAAclgMC3MAQUaHg3wEHYFYO1I-awrk8LPcY2fdjYRcUPA0gigQ-ZBOX8LF7DgKFizhTLPS6CRO3UCwMYuQMDXLN1y0R4pnQriUtwMB6FAfZ0LcHg8xEyx6kYYisB4MXUHFjlZ8AmXxTADi09rW8ENnN-sXfDjALJmblnARcl58wBBEsbrQlCDETAQhbDkjgJSBgfF8TqWdjPZMqZtAEAcgSLQVs0SXDoM9ag2BzDSGMLweAN5pD3R3jvfeCIFhBBDF3fUwBYEECRiZHQUgAD8pd5ZKAnh4Jws9UyH0EOrGugdmFf1oTjT4Hh6EiMJiGEA39u5BC5rWP41BBxQBoRIn-Ns5ba3NPIK0w1Cr5k4TACe59tZFwsaxOwrdlFSk0TjVuyVCid31K3ROFYThRxSImPh888DYHoJnFcIVsYEO4BqP2ZJ4DcHpNrP40TASI0YDSFxKE4DmGwFGIgkJDxSAAAa8IQfPApUh37YC0FiC8jFlgj35KsYpc9XCCOkDQXBVU9hSAANR7CsNyLk_hoH9RXNAxhtVDxYJxHiIpSZmllPIISIQwFWBQAYF8XmLgQHuWMHgIMXdHE0KjiAOwXAlpoDwDoAgmAoCkFEJQGg9AmCIBYGwTgPB-CCGEIgEACcIBJwZBvEABcPAVg8f8isJyznuVQIQYgdzyBUFoIwZgrB2BcDOF8ggIh5jRiWCsdYmxXw1E4i0HY-xNi9hcMcOY4hoyyHjJ0AQZYkQMmAlUmAuRJoQAJFc9kh4QHNG2K4XYHQ_ACG6AcPoMRBjxBpakOl0gGVzMQQSgAohgAZQrWh7DwEENAWBcAwHlcCFIwIpAAE1eDmBjEYb25JpAGEWQarVNB3AgBdUauFqAPB4DNUEFIFkCDEs4vnUBg0WT5kZeKggLK2VYKhCA7OO0uU8r5QKq-bitH5g1Ya2AQIs2fFbtG_wca14VmdICCJ5htlVVqmq2AtaKAqk7PsXgPxHS6nDFCkAQT7wJTwHKOA_AEUPORc8156KPkCH8CId4IA-wincqcDECRSDtB8DGyVlKZX8HiB4RAHgAB6ABOPAvpz0eGZMCzYB7j3igAOxXo8FKqIBDMB3pAEex9z6b2HEiHAOJEB6hrsPSAHkjcPB2DsCc6DQA&query=file%3D%2Fexample.tsx)

## Actual Behavior
Combobox selects the first option in dropdown when <key>space</key> is pressed.

## Expected Behavior
Combobox should allow user to type a space in the input field.

### Logs

_No response_

### Requested priority

Normal

### Products/sites affected

_No response_

### Are you willing to submit a PR to fix?

yes

### Validations

- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.
- [X] The provided reproduction is a minimal reproducible example of the bug.

### PR DESCRIPTION
**Pull request by smhigley**
## Previous Behavior

Out of the box, the space key always selected the current option.

## New Behavior

Only for freeform comboboxes, a separate typing/not typing state is tracked, and when the user is typing the space key inserts a character instead of selecting an option. The conditions for this to happen are:
- If the freeform combobox is closed, space always inserts a character
- If the user's previous keyboard interaction was to insert characters into the text input, space while also insert a character
- If the user has typed, and not used up/down/page up/page down/home/end, or opened the combobox with enter or alt+down since the last typing interaction.

An example: if the user types "hello" and then space, the space character will be inserted. If the user types "hello" and then moves the text insertion character to "h" and presses space, the space character will also be inserted (since they didn't open the dropdown or navigate options since they last typed).

If the user types "hello" and then arrows through a few options and then presses space, the space key will select the highlighted option.

Non-freeform comboboxes do not change behavior and always use space to select.

## Related Issue(s)

- Fixes #26295

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

| Package & Exports                                                                                                             |     Baseline (minified/GZIP) |                           PR |                                                                                                                                                                                                   Change |
| :---------------------------------------------------------------------------------------------------------------------------- | ---------------------------: | ---------------------------: | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: |
| <samp>react-combobox</samp> <br /> <abbr title='bundle-size/Combobox.fixture.js'>Combobox (including child components)</abbr> | `85.509 kB`<br />`27.437 kB` | `85.791 kB`<br />`27.543 kB` | `282 B` <br />`106 B`  |
| <samp>react-combobox</samp> <br /> <abbr title='bundle-size/ComboboxField.fixture.js'>ComboboxField</abbr>                    | `81.926 kB`<br />`27.213 kB` | `82.208 kB`<br />`27.328 kB` | `282 B` <br />`115 B`  |

<details>
<summary>Unchanged fixtures</summary>

| Package & Exports                                                                                                                                                             |          Size (minified/GZIP) |
| ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------: |
| <samp>react-combobox</samp> <br /> <abbr title='bundle-size/Dropdown.fixture.js'>Dropdown (including child components)</abbr>                                                 |  `84.196 kB`<br />`27.191 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ButtonProviderAndTheme.fixture.js'>react-components: Button, FluentProvider & webLightTheme</abbr>              |   `64.09 kB`<br />`17.544 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/MultipleComponents.fixture.js'>react-components: Accordion, Button, FluentProvider, Image, Menu, Popover</abbr> | `202.311 kB`<br />`56.366 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ProviderAndTheme.fixture.js'>react-components: FluentProvider & webLightTheme</abbr>                            |  `35.049 kB`<br />`11.533 kB` |
| <samp>react-portal-compat</samp> <br /> <abbr title='bundle-size/PortalCompat.fixture.js'>PortalCompatProvider</abbr>                                                         |    `6.324 kB`<br />`2.129 kB` |

</details>
<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/cb0acaa586d84b70a9d2c3a9f4a14d5cdd3022cb'>cb0acaa586d84b70a9d2c3a9f4a14d5cdd3022cb</a></sub>
<div id="bundleSizeComment9423"/>

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)
<p>No significant results to display.</p><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27025/perf-test-react-components/Avatar-mount_base.html">1279</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27025/perf-test-react-components/Avatar-mount.html">1276</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27025/perf-test-react-components/Button-mount_base.html">911</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27025/perf-test-react-components/Button-mount.html">934</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Field</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27025/perf-test-react-components/Field-mount_base.html">1957</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27025/perf-test-react-components/Field-mount.html">1932</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27025/perf-test-react-components/FluentProvider-mount_base.html">1544</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27025/perf-test-react-components/FluentProvider-mount.html">1506</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27025/perf-test-react-components/FluentProviderWithTheme-mount_base.html">576</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27025/perf-test-react-components/FluentProviderWithTheme-mount.html">586</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27025/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">540</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27025/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">549</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27025/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">576</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27025/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">591</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>InfoButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27025/perf-test-react-components/InfoButton-mount_base.html">508</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27025/perf-test-react-components/InfoButton-mount.html">517</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27025/perf-test-react-components/MakeStyles-mount_base.html">1952</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27025/perf-test-react-components/MakeStyles-mount.html">1971</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>Persona</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27025/perf-test-react-components/Persona-mount_base.html">2865</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27025/perf-test-react-components/Persona-mount.html">2815</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27025/perf-test-react-components/SpinButton-mount_base.html">2287</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/27025/perf-test-react-components/SpinButton-mount.html">2290</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

**Comment by fabricteam** [comment]
## 🕵 fluentuiv9 No visual regressions between this PR and main
<div id="vrtCommentfluentuiv9"/>

File: packages/react-components/react-combobox/src/components/Combobox/useCombobox.tsx [hunk 1]
```diff
@@ -226,6 +215,49 @@ export const useCombobox_unstable = (props: ComboboxProps, ref: React.Ref<HTMLIn
 
   state.root.ref = useMergedRefs(state.root.ref, rootRef);
 
+  /* Set input.onKeyDown here, so we can override the default behavior for spacebar */
+  const defaultOnTriggerKeyDown = state.input.onKeyDown;
+  state.input.onKeyDown = useEventCallback((ev: React.KeyboardEvent<HTMLInputElement>) => {
```
Comment (GeoffCoxMSFT) [inline packages/react-components/react-combobox/src/components/Combobox/useCombobox.tsx:220]: I just fixed a bug in Tab where I wasn't merging the onClick callback coming in from props with the handler I was setting.  Do you need to merge the callbacks for onKeyDown here? Or does this component not expose the input to callers?

**Review by GeoffCoxMSFT — APPROVED** [review]

File: packages/react-components/react-combobox/src/components/Combobox/useCombobox.tsx (hunk 1 above)
Comment (smhigley) [reply packages/react-components/react-combobox/src/components/Combobox/useCombobox.tsx:220]: The props handler is already merged into `state.input.onKeyDown`, so I'm conditionally calling specifically the props handler vs. the merged handler w/ the selection logic in lines 251-158.

Let me know if that makes sense to you!

**Review by spmonahan — APPROVED** [review]

**Pull request merged**